lib/lists.nix

```diff
@@ -612,6 +616,42 @@ rec {
       if len < 2 then list
       else (sort strictLess pivot.left) ++  [ first ] ++  (sort strictLess pivot.right));
 
+  /*
+    Sort a list based on the default comparison of a derived property `b`.
+
+    This calls the function only once per element, so if your function is non-trivial,
+    it is easily twice as efficient as an unprepared [`sort`](#function-library-lib.lists.sort), and then a log(n) factor.
+
+    However, if the passed function is very cheap, `sort` may be cheaper,
+    because `sort` does not allocate an intermediate list of `(a, b)` pairs.
+
+    The items are returned in `b`-increasing order.
+
+    **Laws**:
+    ```nix
+    sortOn f == sort (a: b: f a < f b)
+    ```
```

Comment:
Needs some better names to not be confusing, `a` and `b` are used for both the arguments and the input/output types of `f`.